File: AiDA/actions.py
import ida_kernwin
import ida_funcs
import ida_name
import ida_hexrays
import re

from . import ui
from .core import ida_utils
from .core.settings import SETTINGS
from .patterns import unreal

class ActionHandler(ida_kernwin.action_handler_t):
    def __init__(self, action_func, plugin_instance):
        super().__init__()
        self.action_func = action_func
        self.plugin = plugin_instance

    def activate(self, ctx):
        self.action_func(ctx, self.plugin)
        return 1

    def update(self, ctx):
        if not self.plugin.ai_client or not self.plugin.ai_client.is_available():
            if self.action_func in [handle_show_settings, handle_scan_for_offsets]:
                return ida_kernwin.AST_ENABLE_FOR_WIDGET
            return ida_kernwin.AST_DISABLE_FOR_WIDGET

        if self.action_func not in [handle_show_settings, handle_scan_for_offsets]:
            if ctx.widget_type not in [ida_kernwin.BWN_PSEUDOCODE, ida_kernwin.BWN_DISASM]:
                return ida_kernwin.AST_DISABLE_FOR_WIDGET

        return ida_kernwin.AST_ENABLE_FOR_WIDGET

def handle_analyze_function(ctx, plugin):
    analysis = plugin.ai_client.analyze_function(ctx.cur_ea)
    if analysis and "Error:" not in analysis:
        ui.show_text_in_viewer(f"AI Analysis for 0x{ctx.cur_ea:X}", analysis)

def handle_rename_function(ctx, plugin):
    func = ida_funcs.get_func(ctx.cur_ea)
    if not func: return
    name = plugin.ai_client.suggest_name(func.start_ea)
    if name and "Error:" not in name:
        if ida_kernwin.ask_yn(ida_kernwin.ASKBTN_YES, f"Rename function at 0x{func.start_ea:X} to:\n\n{name}\n\nApply?") == ida_kernwin.ASKBTN_YES:
            ida_name.set_name(func.start_ea, name, ida_name.SN_CHECK)

def handle_auto_comment(ctx, plugin):
    func = ida_funcs.get_func(ctx.cur_ea)
    if not func: return
    analysis = plugin.ai_client.analyze_function(func.start_ea)
    if analysis and "Error:" not in analysis:
        summary_match = re.search(r"High-Level Purpose:\s*(.*)", analysis)
        summary = summary_match.group(1) if summary_match else analysis.split('\n')[0]
        comment = ida_funcs.get_func_cmt(func, True) or ""
        if "AI Assist:" not in comment:
            new_comment = f"// AI Assist: {summary}\n{comment}"
            ida_funcs.set_func_cmt(func, new_comment, True)
            ida_kernwin.refresh_idaview_anyway()
            vdui = ida_hexrays.get_widget_vdui(ctx.widget)
            if vdui: vdui.refresh_view(True)
            ida_kernwin.msg(f"Comment added to function at 0x{func.start_ea:X}.\n")
        else:
            ida_kernwin.msg("AI-generated comment already exists.\n")

def handle_generate_struct(ctx, plugin):
    from .ai import prompts
    context = ida_utils.get_context_for_prompt(ctx.cur_ea, include_struct_context=True)
    if not context["ok"]:
        ida_kernwin.warning(f"AiDA: {context['message']}")
        return

    prompt = prompts.GENERATE_STRUCT_PROMPT.format(
        code=context["code"],
        struct_context=context["struct_context"]
    )
    struct_cpp = plugin.ai_client._generate(prompt, 0.0)
    if struct_cpp and "Error:" not in struct_cpp:
        ida_utils.apply_struct_from_cpp(struct_cpp, ctx.cur_ea)

def handle_generate_hook(ctx, plugin):
    func = ida_funcs.get_func(ctx.cur_ea)
    if not func: return
    hook_code = plugin.ai_client.generate_hook(func.start_ea)
    if hook_code and "Error:" not in hook_code:
        ui.show_text_in_viewer(f"MinHook Snippet for {ida_name.get_name(func.start_ea)}", hook_code)

def handle_custom_query(ctx, plugin):
    question = ida_kernwin.ask_str("", ida_kernwin.HIST_SRCH, "Ask AI about this function:")
    if question:
        analysis = plugin.ai_client.custom_query(ctx.cur_ea, question)
        if analysis and "Error:" not in analysis:
            ui.show_text_in_viewer(f"AI Query: {question}", analysis)

def handle_scan_for_offsets(ctx, plugin):
    ida_kernwin.msg("====================================================\n")
    ida_kernwin.msg("--- Starting Unreal Engine Pointer Scan ---\n")
    unreal.scan_for_unreal_patterns(plugin.ai_client, SETTINGS)

def handle_show_settings(ctx, plugin):
    ui.SettingsForm.show_and_apply(plugin)

File: actions.cpp
#include "aida_pro.hpp"

int idaapi action_handler::activate(action_activation_ctx_t* ctx)
{
    action_func(ctx, plugin);
    return 1;
}

action_state_t idaapi action_handler::update(action_update_ctx_t* ctx)
{
    if (action_func == handle_show_settings || action_func == handle_scan_for_offsets)
        return AST_ENABLE_ALWAYS;

    if (!plugin || !plugin->ai_client || !plugin->ai_client->is_available())
        return AST_DISABLE_FOR_WIDGET;

    if (ctx->widget_type != BWN_PSEUDOCODE && ctx->widget_type != BWN_DISASM)
        return AST_DISABLE_FOR_WIDGET;

    if (action_func != handle_show_settings && action_func != handle_scan_for_offsets)
    {
        if (get_func(ctx->cur_ea) == nullptr)
            return AST_DISABLE_FOR_WIDGET;
    }

    return AST_ENABLE_FOR_WIDGET;
}

void handle_analyze_function(action_activation_ctx_t* ctx, aida_plugin_t* plugin)
{
    const ea_t func_ea = ctx->cur_ea;
    if (get_func(func_ea) == nullptr)
    {
        warning("AiDA: Please place the cursor inside a function.");
        return;
    }

    auto on_complete = [func_ea](const std::string& analysis) {
        if (!analysis.empty() && analysis.find("Error:") == std::string::npos)
        {
            qstring title;
            title.sprnt("AI Analysis for 0x%a", func_ea);
            show_text_in_viewer(title.c_str(), analysis);
        }
        else if (!analysis.empty())
        {
            warning("AiDA: %s", analysis.c_str());
        }
        };
    plugin->ai_client->analyze_function(func_ea, on_complete);
}

void handle_rename_function(action_activation_ctx_t* ctx, aida_plugin_t* plugin)
{
    func_t* pfn = get_func(ctx->cur_ea);
    if (!pfn)
    {
        warning("AiDA: Please place the cursor inside a function.");
        return;
    }
    const ea_t func_ea = pfn->start_ea;

    qstring current_name;
    if (get_name(&current_name, func_ea) > 0 && is_uname(current_name.c_str()))
    {
        qstring question;
        question.sprnt("HIDECANCEL\nThis function already has a user-defined name ('%s').\nDo you want to ask the AI for a new one anyway?", current_name.c_str());
        if (ask_buttons("~Y~es", "~N~o", nullptr, ASKBTN_NO, question.c_str()) != ASKBTN_YES)
        {
            return;
        }
    }

    auto on_complete = [func_ea](const std::string& name) {
        if (!name.empty() && name.find("Error:") == std::string::npos)
        {
            func_t* pfn_cb = get_func(func_ea);
            if (!pfn_cb)
            {
                warning("AiDA: Function at 0x%a no longer exists.", func_ea);
                return;
            }

            qstring clean_name = name.c_str();
            clean_name.replace("`", "");
            clean_name.trim2();

            if (!validate_name(&clean_name, VNT_IDENT))
            {
                warning("AiDA: The suggested name '%s' is not a valid identifier.", clean_name.c_str());
                return;
            }

            qstring question;
            question.sprnt("Rename function at 0x%a to:\n\n%s\n\nApply this change?", pfn_cb->start_ea, clean_name.c_str());
            if (ask_buttons("~Y~es", "~N~o", nullptr, ASKBTN_YES, question.c_str()) == ASKBTN_YES)
            {
                if (set_name(pfn_cb->start_ea, clean_name.c_str(), SN_FORCE))
                {
                    msg("AiDA: Function at 0x%a renamed to '%s'.\n", pfn_cb->start_ea, clean_name.c_str());
                }
                else
                {
                    warning("AiDA: Failed to set new function name. It might be invalid or already in use.");
                }
            }
        }
        else if (!name.empty())
        {
            warning("AiDA: %s", name.c_str());
        }
        };
    plugin->ai_client->suggest_name(func_ea, on_complete);
}

void handle_auto_comment(action_activation_ctx_t* ctx, aida_plugin_t* plugin)
{
    func_t* pfn = get_func(ctx->cur_ea);
    if (!pfn)
    {
        warning("AiDA: Please place the cursor inside a function.");
        return;
    }
    const ea_t func_ea = pfn->start_ea;

    auto on_complete = [func_ea](const std::string& analysis) {
        if (!analysis.empty() && analysis.find("Error:") == std::string::npos)
        {
            func_t* pfn_cb = get_func(func_ea);
            if (!pfn_cb)
            {
                warning("AiDA: Function at 0x%a no longer exists.", func_ea);
                return;
            }

            std::smatch match;
            std::string summary;
            if (std::regex_search(analysis, match, std::regex("High-Level Purpose:\\s*(.*)")))
            {
                summary = match[1].str();
            }
            else
            {
                size_t first_line_end = analysis.find('\n');
                summary = analysis.substr(0, first_line_end);
            }

            qstring comment;
            get_func_cmt(&comment, pfn_cb, true);
            if (comment.find("AI Assist:") == qstring::npos)
            {
                qstring new_comment;
                if (comment.empty())
                    new_comment.sprnt("// AI Assist: %s", summary.c_str());
                else
                    new_comment.sprnt("// AI Assist: %s\n%s", summary.c_str(), comment.c_str());

                set_func_cmt(pfn_cb, new_comment.c_str(), true);

                if (init_hexrays_plugin())
                {
                    cfuncptr_t cfunc = decompile(pfn_cb);
                    if (cfunc != nullptr)
                    {
                        cfunc->refresh_func_ctext();
                    }
                    else
                    {
                        mark_cfunc_dirty(pfn_cb->start_ea, true);
                        request_refresh(IWID_DISASM);
                    }
                }
                else
                {
                    request_refresh(IWID_DISASM);
                }

                msg("AiDA: Comment added to function at 0x%a.\n", pfn_cb->start_ea);
            }
            else
            {
                msg("AiDA: AI-generated comment already exists.\n");
            }
        }
        else if (!analysis.empty())
        {
            warning("AiDA: %s", analysis.c_str());
        }
    };
    plugin->ai_client->analyze_function(func_ea, on_complete);
}

void handle_generate_struct(action_activation_ctx_t* ctx, aida_plugin_t* plugin)
{
    const ea_t func_ea = ctx->cur_ea;
    if (get_func(func_ea) == nullptr)
    {
        warning("AiDA: Please place the cursor inside a function to generate a struct.");
        return;
    }

    auto on_complete = [func_ea](const std::string& struct_cpp) {
        if (!struct_cpp.empty() && struct_cpp.find("Error:") == std::string::npos)
        {
            ida_utils::apply_struct_from_cpp(struct_cpp, func_ea);
        }
        else if (!struct_cpp.empty())
        {
            warning("AiDA: %s", struct_cpp.c_str());
        }
        };
    plugin->ai_client->generate_struct(func_ea, on_complete);
}

void handle_generate_hook(action_activation_ctx_t* ctx, aida_plugin_t* plugin)
{
    func_t* pfn = get_func(ctx->cur_ea);
    if (!pfn)
    {
        warning("AiDA: Please place the cursor inside a function to generate a hook.");
        return;
    }
    const ea_t func_ea = pfn->start_ea;

    auto on_complete = [func_ea](const std::string& hook_code) {
        if (!hook_code.empty() && hook_code.find("Error:") == std::string::npos)
        {
            qstring func_name;
            get_func_name(&func_name, func_ea);
            qstring title;
            title.sprnt("MinHook Snippet for %s", func_name.c_str());
            show_text_in_viewer(title.c_str(), hook_code);
        }
        else if (!hook_code.empty())
        {
            warning("AiDA: %s", hook_code.c_str());
        }
        };
    plugin->ai_client->generate_hook(func_ea, on_complete);
}

void handle_custom_query(action_activation_ctx_t* ctx, aida_plugin_t* plugin)
{
    const ea_t func_ea = ctx->cur_ea;
    if (get_func(func_ea) == nullptr)
    {
        warning("AiDA: Please place the cursor inside a function for a custom query.");
        return;
    }

    qstring question;
    if (ask_str(&question, HIST_SRCH, "Ask AI about this function:"))
    {
        auto on_complete = [question](const std::string& analysis) {
            if (!analysis.empty() && analysis.find("Error:") == std::string::npos)
            {
                qstring title;
                title.sprnt("AI Query: %s", question.c_str());
                show_text_in_viewer(title.c_str(), analysis);
            }
            else if (!analysis.empty())
            {
                warning("AiDA: %s", analysis.c_str());
            }
            };
        plugin->ai_client->custom_query(func_ea, question.c_str(), on_complete);
    }
}

void handle_scan_for_offsets(action_activation_ctx_t* /*ctx*/, aida_plugin_t* /*plugin*/)
{
    msg("====================================================\n");
    msg("--- Starting Unreal Engine Pointer Scan ---\n");
    warning("Scan for Engine Pointers is not yet implemented in the C++ version.");
    // unreal::scan_for_unreal_patterns(plugin->ai_client, g_settings); COMING SOON!!!
}

void handle_show_settings(action_activation_ctx_t* /*ctx*/, aida_plugin_t* plugin)
{
    SettingsForm::show_and_apply(plugin);
}